modules/net-vpc-firewall/main.tf

```diff
@@ -15,149 +15,107 @@
-
-resource "google_compute_firewall" "allow-tag-ssh" {
-  count         = length(var.ssh_source_ranges) > 0 ? 1 : 0
-  name          = "${var.network}-ingress-tag-ssh"
-  description   = "Allow SSH to machines with the 'ssh' tag"
-  network       = var.network
-  project       = var.project_id
-  source_ranges = var.ssh_source_ranges
-  target_tags   = ["ssh"]
-  allow {
-    protocol = "tcp"
-    ports    = ["22"]
-  }
-}
-
-resource "google_compute_firewall" "allow-tag-http" {
-  count         = length(var.http_source_ranges) > 0 ? 1 : 0
-  name          = "${var.network}-ingress-tag-http"
-  description   = "Allow HTTP to machines with the 'http-server' tag"
-  network       = var.network
-  project       = var.project_id
-  source_ranges = var.http_source_ranges
-  target_tags   = ["http-server"]
-  allow {
-    protocol = "tcp"
-    ports    = ["80"]
+    for r in local._factory_rule_list : r.name => r
   }
-}
-
-resource "google_compute_firewall" "allow-tag-https" {
-  count         = length(var.https_source_ranges) > 0 ? 1 : 0
-  name          = "${var.network}-ingress-tag-https"
-  description   = "Allow HTTPS to machines with the 'https' tag"
-  network       = var.network
-  project       = var.project_id
-  source_ranges = var.https_source_ranges
-  target_tags   = ["https-server"]
-  allow {
-    protocol = "tcp"
-    ports    = ["443"]
+  _named_ranges = merge(
+    var.named_ranges,
+    try(yamldecode(file(var.factories_config.cidr_tpl_file)), {})
+  )
+  # convert rules data to resource format and replace range template variables
+  rules = {
+    for name, rule in merge(local._factory_rules, var.custom_rules) :
+    name => merge(rule, {
+      action    = rule.is_deny == true ? "DENY" : "ALLOW"
+      direction = rule.is_egress == true ? "EGRESS" : "INGRESS"
+      rules     = { for k, v in rule.rules : k => v }
+      ranges = flatten([
+        for range in(rule.ranges == null ? [] : rule.ranges) :
+        try(local._named_ranges[range], range)
+      ])
+    })
   }
 }
 
-################################################################################
-#                                dynamic rules                                 #
-################################################################################
-
 resource "google_compute_firewall" "custom-rules" {
-  # provider                = "google-beta"
-  for_each    = local.custom_rules
+  for_each    = local.rules
+  project     = var.project_id
+  network     = var.network
   name        = each.key
   description = each.value.description
   direction   = each.value.direction
-  network     = var.network
-  project     = var.project_id
   source_ranges = (
     each.value.direction == "INGRESS"
-    ? coalesce(each.value.ranges, []) == [] ? ["0.0.0.0/0"] : each.value.ranges
-    : null
+    ? (
+      coalesce(each.value.ranges, []) == []
+      ? ["0.0.0.0/0"]
+      : each.value.ranges
```

Comment:
I think the whole inner ternary can be replaced with just `coalesce(each.value.ranges, ["0.0.0.0/0"])`

Reply:
you don't catch the empty list in that way